[opentitanlib] Refactor lifecycle controller utilities to use bitflags

jwnrt opened the pull request:
## Summary

These changes:

* Update the [bitflags] dependency from v1.0 to v2.1.
* Replace the `LcBit` trait with `bitflags` types.

This allows us to use [bitflags methods](https://docs.rs/bitflags/2.1.0/bitflags/example_generated/struct.Flags.html) for things like checking if bits are set, making masks from unions of bits, etc.

[bitflags]: https://crates.io/crates/bitflags

dmcardle commented on sw/host/opentitanlib/src/test_utils/lc_transition.rs:
```diff
@@ -47,11 +48,10 @@ pub fn trigger_lc_transition(
-        return Err(LcTransitionError::LcCtrlNotReady(lc_status).into());
+    let status = jtag.read_lc_ctrl_reg(&LcCtrlReg::Status)?;
+    let status = LcCtrlStatus::from_bits(status).ok_or(LcTransitionError::InvalidState(status))?;
+    if !status.contains(LcCtrlStatus::INITIALIZED | LcCtrlStatus::READY) {
+        return Err(LcTransitionError::LcCtrlNotReady(status).into());
     }
```
It looks like this is a behavior change from the prior code. It used to return an error if, for instance, the STATUS register were `INITIALIZED | READY | TOKEN_ERROR`. I think we want a regular equality check here rather than `contains()`.

dmcardle commented on sw/host/opentitanlib/src/test_utils/lc_transition.rs:
```diff
@@ -90,31 +90,27 @@ pub fn trigger_lc_transition(
-            != 0
-        {
+
+        if !status.contains(
+            LcCtrlStatus::INITIALIZED | LcCtrlStatus::READY | LcCtrlStatus::TRANSITION_SUCCESSFUL,
```
I think this is a behavior change. The code used to return an error when STATUS contained any flags other than INITIALIZED, READY, or TRANSITION_SUCCESSFUL. Now it returns an error if status is missing one of those flags.

To preserve the original behavior, this should be something like this:

```rust
let expected_status = LcCtrlStatus::INITIALIZED | LcCtrlStatus::READY | LcCtrlStatus::TRANSITION_SUCCESSFUL;
if status != expected_status {
    return Err(...);
}
```

jwnrt replied:
Thanks, as above I've switched back to `!=` for these and will add a separate `ERRORS` check in the future if we need it.

dmcardle commented on third_party/rust/crates/BUILD.bitflags-2.1.0.bazel:
```diff
@@ -0,0 +1,47 @@
+###############################################################################
```
If this PR eliminates dependencies on older versions of bitflags, let's also delete BUILD.bitflags-1.3.2.bazel andBUILD.bitvec-1.0.1.bazel!

jwnrt replied:
As above, these are transient dependencies that can't be removed.

jwnrt commented on third_party/rust/crates/defs.bzl:
```diff
@@ -604,6 +604,16 @@ def crate_repositories():
 
+    maybe(
+        http_archive,
+        name = "crate_index__bitflags-2.1.0",
```
Sure, I'll have a go but I'm not very comfortable with the Bazel Rust stuff

jwnrt replied:
I'm not sure how to check without `cargo tree`, but I think `bitflags-1.3.2` is a transient dependency of other dependencies, which is why it's still vendored in by Bazel. In that case, we can't remove these crate defs.

jwnrt commented on sw/host/opentitanlib/src/dif/lc_ctrl.rs:
```diff
@@ -119,57 +122,58 @@ impl LcCtrlReg {
+        const BUS_INTEG_ERROR        = 0b1 << bindgen::dif::LC_CTRL_STATUS_BUS_INTEG_ERROR_BIT;
+        const OTP_PARTITION_ERROR    = 0b1 << bindgen::dif::LC_CTRL_STATUS_OTP_PARTITION_ERROR_BIT;
+
+        const ERROR =
+            Self::TRANSITION_COUNT_ERROR.bits() |
+            Self::TRANSITION_ERROR.bits() |
+            Self::TOKEN_ERROR.bits() |
+            Self::FLASH_RMA_ERROR.bits() |
+            Self::OTP_ERROR.bits() |
+            Self::STATE_ERROR.bits() |
+            Self::BUS_INTEG_ERROR.bits() |
+            Self::OTP_PARTITION_ERROR.bits();
```
I'm using this in a WIP PR, but I can re-add that when I need it.

jwnrt replied:
Removed in the fixup, will add this back when I need it because the mask of errors is useful!

timothytrippel replied:
Just an FYI, if you are using something in a WIP PR, you don't need to delete work just to redo it again. It's OK to be efficient :)

jwnrt replied:
Thanks Tim, will do next time. I'm happy to remove here since it's small

jwnrt commented on sw/host/opentitanlib/src/dif/lc_ctrl.rs:
```diff
@@ -212,14 +216,4 @@ mod tests {
     }
-
-    #[test]
-    fn lc_bit_union() {
```
Added something similar to the previous test. I ran `bazel test //sw/host/opentitanlib` but not sure if that actually tested it.

jwnrt commented on third_party/rust/Cargo.lock:
```diff
@@ -136,6 +136,15 @@ version = "1.3.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "bef38d45163c2f1dde094a7dfd33ccf595c92905c8f8f4fdc18d06fb1037718a"
 
+[[package]]
```
See above, these look like transient dependencies so can't be removed.

jwnrt commented:
Squashed fixup which resolves review comments:

* Added a test showing usage of the bitfields.
* Restored behaviour change from checking if status `contains` bits to comparing the full status.

jwnrt commented:
Rebased to (hopefully) fix CI which I couldn't work out why it was failing.

jwnrt commented:
Nope, it was a real error after all (FYI it's not `bazel test //sw/host/opentitanlib`, it's `bazel test //sw/host/opentitanlib:opentitanlib_test`. Fixed.

jwnrt commented:
Rebased to fix conflicts with https://github.com/lowRISC/opentitan/pull/17996.

Rerun `bazel test //sw/host/opentitanlib:opentitanlib_test` and seems fine.
